src/auth/discord.rs

```diff
@@ -1 +1,404 @@
+// TODO: This has be set while creating the Discord OAuth Application
+pub const CALLBACK_URL: &str = "/auth/discord/callback/";
+pub const AUTH_URL: &str = "https://discord.com/oauth2/authorize";
+pub const TOKEN_URL: &str = "https://discord.com/api/oauth2/token";
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct UserDetail {
+    pub token: String,
+    pub user_name: String,
+    pub user_id: String,
+}
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct DiscordAuthReq {
+    pub client_secret: String,
+    pub client_id: String,
+    pub grant_type: String,
+    pub code: String,
+    pub redirect_uri: String,
+}
+pub(crate) enum DiscordScopes {
+    Identify,
+    Guilds,
+    GuildsMembersRead,
+}
 
+impl DiscordScopes {
+    pub(crate) fn as_str(&self) -> &'static str {
+        match self {
+            DiscordScopes::Identify => "identify",
+            DiscordScopes::Guilds => "guilds",
+            DiscordScopes::GuildsMembersRead => "guilds.members.read",
+        }
+    }
+}
+// route: /auth/login/
+pub async fn login(req: actix_web::HttpRequest) -> fpm::Result<fpm::http::Response> {
+    // Discord will be redirect to this url after login process completed
+
+    let redirect_url: String = format!(
+        "{}://{}{}",
+        req.connection_info().scheme(),
+        req.connection_info().host(),
+        CALLBACK_URL
+    );
+    let client_id = match std::env::var("DISCORD_CLIENT_ID") {
+        Ok(id) => id,
+        Err(_e) => {
+            println!("WARN: DISCORD_CLIENT_ID not set");
+            // TODO: Need to change this approach later
+            "FPM_TEMP_DISCORD_CLIENT_ID".to_string()
```

Comment:
Return error if discord client ID is not found.

Reply:
done